api/src/main/java/org/openmrs/module/pihcore/htmlformentry/action/CloseCovidProgramAction.java

```diff
@@ -0,0 +1,112 @@
+
+    // TODO if we decide we want to use this pattern to other programs, would make sense to see if we can pass in
+    // TODO the program uuid to this action; would probably mean expanding the PostSubmissionAction tag to accept parameters
+
+    protected Log log = LogFactory.getLog(getClass());
+
+    private static String COVID_PROGRAM_UUID = "dc227c44-d68e-431b-a20d-a60bb3496543";
+
+    private static String DISPOSITION_TRANSFER_OUT = "transferOutOfHospital";
+
+    private static String DISPOSITION_DEATH = "markPatientDead";
+
+    private static String DISPOSITION_DISCHARGE = "discharge";
+
+    private static String CONCEPT_TRANSFER_OUT = "PATIENT TRANSFERRED OUT";
+
+    private static String CONCEPT_DIED = "PATIENT DIED";
+
+    private static String CONCEPT_DISCHARGE = "TREATMENT STOPPED";
+
+    @Override
+    public void applyAction(FormEntrySession formEntrySession) {
+        closeCovidProgram(formEntrySession, Context.getProgramWorkflowService(), Context.getService(DispositionService.class), Context.getConceptService());
+    }
+
+
+    private void closeCovidProgram(FormEntrySession formEntrySession, ProgramWorkflowService programWorkflowService, DispositionService dispositionService, ConceptService conceptService) {
+        Patient patient = formEntrySession.getPatient();
+        Program covid = programWorkflowService.getProgramByUuid(COVID_PROGRAM_UUID);
+        Encounter encounter = formEntrySession.getEncounter();
+        // use the encounter date *without* the time component
+        Date encounterDate = (new DateTime(encounter.getEncounterDatetime())).withTimeAtStartOfDay().toDate();
+
+        // we will update any COVID program started on or before the encounter date AND EITHER closed on or after encounter date OR not closed at all
+        // note that this should handle moving a completion date *earlier* but *not later*:
+        // ie if a patient is enrolled on Jan 11 and program is completed on Jan 14 and discharge form is entered on Jan 12, we assume
+        // Jan 12 is the correct completion date and update
+        // however, if a discharge form is entered on Jan 16, the program from Jan 11 to Jan 14 is *not* updated
+
+        List<PatientProgram> candidates = programWorkflowService.getPatientPrograms(patient, covid, null,
+                encounterDate, encounterDate, null, false);
+
+        if (candidates != null) {
+
+            if (candidates.size() > 1) {
+                log.warn("More than one COVID program enrolment for patient " + patient.getId() + " on date " + encounterDate);
+            }
+
+            for (PatientProgram patientProgram : candidates) {
+
+                // set the date completed to the encounter date
+                patientProgram.setDateCompleted(encounterDate);
+
+                // find the disposition
+                DispositionDescriptor dispositionDescriptor = dispositionService.getDispositionDescriptor();
+
+                Disposition disposition = null;
```

Comment:
```suggestion
                Disposition disposition;
```

Reply:
Without setting dispositon to null here you get a "variable might not have been initialized" warning on line 90 (which would happen if it never may any trips through the loop that starts on line 82)